### ISSUE TITLE
Mac OS window persists and hangs after destroy window and quit

### ISSUE THREAD

**Issue by rongcuid** [libsdl-org/SDL#10081]
On development SDL3 branch:

```c
#include <unistd.h>
#include <SDL3/SDL.h>

int main() {
	SDL_Init(SDL_INIT_VIDEO);
	SDL_Window * window = SDL_CreateWindow("Test", 800, 600, 0);
	SDL_ShowWindow(window);
	SDL_Event ev;
	while (1) {
		if (SDL_PollEvent(&ev)) {
			if (ev.type == SDL_EVENT_QUIT)
				break;
		}
	}
	SDL_DestroyWindow(window);
	SDL_Quit();
	sleep(10);
	return 0;
}
```

Reproduce:
- Launch program
- Close window

Expected:
- Window closes immediately

Actual:
- Window freezes until the end of program

This is a problem for my project where windows need to be opened/closed repeatedly.

**Comment by madebr** [comment]
This also reproduces on Linux with `testsprite --windows 2`.

**Comment by Kontrabant** [comment]
macOS requires pumping the events after destroying a window for the OS to process the change and remove it from the desktop. This has been the case going back to SDL2.

> This also reproduces on Linux with `testsprite --windows 2`: pressing the close button does nothing.

This is due to the test framework and testsprite not handling SDL_EVENT_WINDOW_CLOSE_REQUESTED (it's printed when event logging is enabled, but doesn't actually do anything). It works with only one window since SDL will just send SDL_EVENT_QUIT in that case.

**Comment by rongcuid** [comment]
> macOS requires pumping the events after destroying a window for the OS to process the change and remove it from the desktop. This has been the case going back to SDL2.

Do you mean the user needs to do that? Is there documentation on this behavior?

**Comment by icculus** [comment]
I added a note to the SDL_DestroyWindow docs; in practice most apps will either pump the event queue regularly in their main loop logic, use the main callbacks, or be about to shut down after destroying their windows, so I don't think this issue needs further attention.

**Comment by hmatheisen** [comment]
Pumping the events after closing the window still does not work on Mac OS.

I tried running the same code and pumping the events:

```C
#include <SDL3/SDL.h>
#include <unistd.h>

int main() {
  SDL_Init(SDL_INIT_VIDEO);
  SDL_Window *window = SDL_CreateWindow("Test", 800, 600, 0);
  SDL_ShowWindow(window);
  SDL_Event ev;
  while (1) {
    if (SDL_PollEvent(&ev)) {
      if (ev.type == SDL_EVENT_QUIT)
        break;
    }
  }
  SDL_DestroyWindow(window);
  SDL_PumpEvents();
  SDL_Quit();
  sleep(10);

  return 0;
}
```

And the window still happens to be stuck.
I am using SDL 3.2.16

**Comment by XavierCS-dev** [comment]
I am on linux wayland, the window closes immediately by SDL_DestroyWindow() hangs for a few seconds

**Comment by pseudo-rnd-thoughts** [comment]
### Tl;dr identified the root cause and https://github.com/libsdl-org/SDL/pull/15992 contains a fix

This issue has annoying me for many years upstream in PyGame and Gymnasium so I used fable to dig into this issue.

Using lldb to debug the root cause, fable found that the implicit windows close/hide animation is problem which helpful explains the why the pumping behaviour fixed the issue in the past.

What actually happens is that -[NSWindow close] (and orderOut:, which SDL hits first via SDL_HideWindow inside SDL_DestroyWindow) doesn't remove the window from the screen. It starts _NSWindowTransformAnimation (via -[NSWindow _doOrderWindow:], ~250ms), and the window-server side of the removal is only committed when the animation completes which requires the app's main runloop to be serviced. If the app destroys its window and then never services the runloop again (Jupyter kernels, REPLs, test harnesses, i.e., anything that keeps the process alive without pumping), the animation freezes at frame zero and the fully-opaque dead window stays on screen until the process exits.

Measured behavior on macOS 26 (but the mechanism, and the reports here, go back years):

| after SDL_DestroyWindow       | result                                             |
|-------------------------------|----------------------------------------------------|
| nothing                       | ghost forever                                      |
| 1× SDL_PumpEvents immediately | ghost (animation starts, never finishes)           |
| 20× pump back-to-back         | ghost (all passes land before the animation ends)  |
| 400ms sleep, then pump        | ghost (animation never started)                    |
| pump → 300ms → pump           | clean (second pass lands after animation duration) |
| continuous runloop servicing  | clean at ~250ms                                    |

This also explains @hmatheisen's report that pumping "doesn't work" as a pump loop shorter than the animation duration, or one that isn't running ~250ms after the destroy, still ghosts. No NSEvent is involved at any point: a bare 10ms CFRunLoopRunInMode pass works wherever pumping works, and blocking in nextEventMatchingMask fixes it with no event ever arriving.

Fix (two lines, tested): set nswindow.animationBehavior = NSWindowAnimationBehaviorNone at window creation (creation-time matters — at destroy time the SDL_HideWindow order-out has already started the animation), and give AppKit one short runloop pass (CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.02, false)) after [nswindow close] in Cocoa_DestroyWindow so the now-non-animated order-out is committed even if the app never pumps again. With this, the window is gone ~20ms after SDL_DestroyWindow with zero pumping, verified A/B in a live Jupyter kernel (the pygame-ce repro from pygame-community/pygame-ce#2256).

PR: https://github.com/libsdl-org/SDL/pull/15992. Happy to port the same two lines to main/SDL3 — the Cocoa backend structure is the same.

**Comment by slouken** [comment]
@icculus, this seems like a good fix, can you investigate for the next release?

**Comment by icculus** [comment]
Yeah, I'll take a look at this later this week.

**Comment by pseudo-rnd-thoughts** [comment]
@icculus Let me know if you want more information / debug scripts used in the process to help in the fix

**Comment by icculus** [comment]
If I apply your patch, it will often work the first time, but if I rerun the reproduction case a second time, it often won't, leaving the window on the screen for the full 10 seconds. I increased the CFRunLoop time from 0.02 to 2.0 and it still didn't work out, so there's something else still going on here that I don't understand yet.

(This is on mac 15.7.4, maybe behavior changed on macOS 26 or whatever.)

Sticking this in the app before the SDL_Quit() call _does_ seem to always fix it:

```c
const Uint64 start = SDL_GetTicks();
while ((SDL_GetTicks() - start) <= 250) {
    SDL_PumpEvents();
}
```

Which I suppose isn't surprising.

Also, if this approach is viable, we could probably check `window->is_destroying` in Cocoa_HideWindow() and set the animation behavior there, to avoid messing with the system's animation whims in general...but also I'm wary of turning this off at all, in case a future macOS wants to do some more dramatic, swoopy animation that users love and that we are breaking through this mechanism in the name of a corner-case.

If nothing else, we can have SDL_Quit() force a small internal pumpevents delay on macOS, to allow windows to clean up. I think that wouldn't be unreasonable, and would handle this reproduction case, even if we don't want to do a delay for each window destruction.

**Comment by slouken** [comment]
> If nothing else, we can have SDL_Quit() force a small internal pumpevents delay on macOS, to allow windows to clean up. I think that wouldn't be unreasonable, and would handle this reproduction case, even if we don't want to do a delay for each window destruction.

We shouldn't actually call pump events, since we don't want to change SDL state or deliver events in SDL_Quit(), but we could run the run loop for a short time, that seems reasonable.

**Comment by pseudo-rnd-thoughts** [comment]
@slouken @icculus Would you like me to update the PR to use the following code instead? 
```
const Uint64 start = SDL_GetTicks();
while ((SDL_GetTicks() - start) <= 250) {
    SDL_PumpEvents();
}
```

**Comment by icculus** [comment]
No thanks, I'm going to explore this more for the next milestone.

**Comment by icculus** [comment]
Okay, this seems to work here, during SDL_Quit()...

```diff
diff --git a/src/video/cocoa/SDL_cocoavideo.m b/src/video/cocoa/SDL_cocoavideo.m
index b0a8b6945..c87a48cd6 100644
--- a/src/video/cocoa/SDL_cocoavideo.m
+++ b/src/video/cocoa/SDL_cocoavideo.m
@@ -235,6 +235,14 @@ void Cocoa_VideoQuit(SDL_VideoDevice *_this)
 {
     @autoreleasepool {
         SDL_CocoaVideoData *data = (__bridge SDL_CocoaVideoData *)_this->internal;
+
+        // Run the CFRunLoop a little, so last-minute window animations can run, etc, but we hopefully don't change any significant state.
+        // This needs to run multiple times over a time period, it's not enough to give it a longer timeout and use "false" for returnAfterSourceHandled.
+        const Uint64 start = SDL_GetTicks();
+        do {
+            CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.0, false);
+        } while ((SDL_GetTicks() - start) <= 20);
+
         Cocoa_QuitModes(_this);
         Cocoa_QuitKeyboard(_this);
         Cocoa_QuitGCMouse();
```

One can also use `[NSApp nextEventMatchingMask:NSEventMaskAny untilDate:[NSDate distantPast] inMode:NSDefaultRunLoopMode dequeue:NO];` instead of `CFRunLoopRunInMode`, but using CoreFoundation here felt like less risk of changing Cocoa state, but I don't know.

Running for 10 milliseconds seemed to reliably fix it here. 5 did not. I went with 20, and we only do this during SDL_Quit() instead of SDL_DestroyWindow(); if the app doesn't quit the subsystem and also doesn't reliably pump the event queue, I consider that an app bug. I'm hoping the need to do this for 250 milliseconds was just Claude hallucinating, but we can adjust as necessary.

Should I commit this?

**Comment by slouken** [comment]
> Should I commit this?

Yes.

**Comment by icculus** [comment]
Done and cherry-picked!

**Comment by FeldrinH** [comment]
Could this also be cherry picked for SDL 2, so it eventually reaches Pygame?
